Refine default Python runtime sequence behavior; remove experimental code.

hunhoffe opened the pull request:
Previously, the default behavior for runtime sequences that don't use task groups was to insert dma_frees after each operation that waited. This can lead to weird behavior, even for simple cases such as two drains in a row, as noted in https://github.com/Xilinx/mlir-aie/issues/2653. 

This PR changes the default behavior to use a default task group when a task group is not specified. This introduces a nice property this ensures runtime sequences using a single task group run the same was as runtime sequences that don't explicitly use a task group. 

I think this is easier to reason about, but would welcome feedback!

The only code that really relied on the default behavior was some experimental code I had added some time back. I've deleted that code; it should be easy enough to fetch in the git history if it is needed in the future.

TODOs:
- [x] Create tests to verify the default behavior
- [x] Modify the default behavior if necessary
- [x] Ensure the default behavior is documented in the programming guide

andrej commented on python/iron/runtime/runtime.py:
```diff
@@ -319,36 +319,36 @@ def sequence(*args):
+                        fn(*args)
+                for fn, args in actions:
+                    if fn != dma_await_task:
+                        fn(*args)
```
Not familiar with the code so please correct me if I'm misunderstanding something. Is case specifically for the the `dma_free_task` action, or are there others here? That would make sense for me (wait first, then free everything). If it's only `free`, can we make it explicit and raise an Error if any other actions other than `await` and `free` are encountered in the list? I think it would make the intent clearer.

hunhoffe commented:
@andrej I quite like your suggestion, and have enabled it be default, but I did leave an escape hatch so "expert users" may mix default/not default if they wish through a flag to the ```Runtime()``` constructor.

I also updated the tests to test this flag. I then found a fair amount of programming examples that mixed default/explicit task groups, and updated them with a more detailed comment.
